### PR TITLE
Make Elasticsearch index URL configurable per pipeline run

### PR DESCRIPTION
**Pull request by valiparsa**
### Description

Make Elasticsearch index URL configurable per pipeline run.

For details see the change log.

### Dependency release notes

N/A

### Submitter checklist

- [x] Change has been tested (on a back-end cluster)
- [x] [changelog.yaml] with upgrade notes are prepared and appropriate for the audience affected by the change (users or developer, depending on the change).
- [x] Semantic version diffed against [last release][releases] and updated accordingly. In this project the version has to be maintained here:
  - [/charts/steward/Chart.yaml](https://github.com/SAP/stewardci-core/blob/master/charts/steward/Chart.yaml) (`version` and `appVersion`)

In case dependencies have been updated:
- n/a Links to external changelogs added since the last release of our component
- n/a Changelogs read thoroughly, potential impact described, upgrade notes prepared (if necessary)
- n/a Check if dependency update affects our semantic version increment.

### Reviewer checklist

Before the changes are marked as `ready-for-merge`: 

- [x] There are at least 2 approvals for the pull request and no outstanding requests for change
- [x] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [x] The Pull Request title is understandable and reflects the changes well
- [x] The Pull Request description is understandable and well documented
- [x] 'Upgrade notes' are documented in changelog.yaml (if required)
- [x] [changelog.yaml] entry for this Pull Request has been added
  - [x] Changelog entry contains all required information

[changelog.yaml]: https://github.com/SAP/stewardci-core/changelog.yaml
[releases]: https://github.com/SAP/stewardci-core/releases

### CODE REVIEW

File: changelog.yaml
Comment (rinckm) [inline changelog.yaml:34]:
```suggestion
```
You should not have two entries with version NEXT but two changes in version NEXT.

File: test/builder/pipelineRun.go
Comment (rinckm) [inline test/builder/pipelineRun.go:151]: You should only set one aspect of the logging here, not both.
Check is logging is nil -> Create it, else modify only one aspect. See Line 105 as example.
Add tests

Comment (rinckm) [reply test/builder/pipelineRun.go:151]: Rename Logging above to LoggingWithRunID

**Review by rinckm — CHANGES_REQUESTED** [review]
See comments inline

File: pkg/runctl/run_manager.go
Comment (rinckm) [inline pkg/runctl/run_manager.go:306]:
```suggestion
```
This is not required as this is no secret to clone from a git repository if I got it right.

File: pkg/runctl/run_manager.go
Comment (rinckm) [inline pkg/runctl/run_manager.go:301]: This is a copy and paste error IMHO.
```
```

File: pkg/runctl/run_manager.go
Comment (rinckm) [inline pkg/runctl/run_manager.go:305]: This line will create a secret in the run namespace with a new unique name.
This name must be used in line 714 rather than the name from the spec.

File: pkg/runctl/run_manager.go
Comment (rinckm) [inline pkg/runctl/run_manager.go:237]:
```suggestion
	ctx.elasticSearchCredentilaName, err = c.copyElasticSearchCredentialToRunNamespace(ctx, secretHelper)
```
Extend the context to be able to hold the elasticSearchCredentilaName

File: pkg/runctl/run_manager.go
Comment (rinckm) [inline pkg/runctl/run_manager.go:714]:
```suggestion
			params = append(params, tektonStringParam("PIPELINE_LOG_ELASTICSEARCH_CREDENTIAL", ctx.elasticSearchCredentilaName))
```

File: changelog.yaml
Comment (alxsap) [inline changelog.yaml:41]:
```suggestion
    title: Make Elasticsearch IndexURL configurable in PipelineRun custom resource objects.

```

File: changelog.yaml
Comment (alxsap) [inline changelog.yaml:44]:
```suggestion
      push logs to different indices in Elasticsearch. This change allows developers to

```

File: changelog.yaml
Comment (alxsap) [inline changelog.yaml:45]:
```suggestion
      make Elasticsearch IndexURL configurable in Steward PipelineRun objects as a parameter.

```

File: charts/steward/templates/clustertask-jenkinsfile-runner.yaml
Comment (alxsap) [inline charts/steward/templates/clustertask-jenkinsfile-runner.yaml:109]: Is it OK with `inputs.`? See https://github.com/SAP/stewardci-core/pull/173

File: pkg/apis/steward/v1alpha1/run_types.go
Comment (alxsap) [inline pkg/apis/steward/v1alpha1/run_types.go:141]:
```suggestion
	ElasticsearchCredential string `json:"elasticsearchCredential,omitempty"`
```
and other occurrences

File: pkg/runctl/run_manager.go
Comment (alxsap) [inline pkg/runctl/run_manager.go:287]: Add docu, especially return parameters

File: pkg/runctl/run_manager.go
Comment (alxsap) [reply pkg/runctl/run_manager.go:310]: No need for `repoServerURL` at all

File: pkg/runctl/run_manager.go
Comment (alxsap) [inline pkg/runctl/run_manager.go:801]:
```suggestion
func ensureValidURL(indexURL string) (string, error) {
```
`toValidURL` sounds like invalid URLs are getting fixed by this function

File: pkg/runctl/run_manager.go
Comment (alxsap) [inline pkg/runctl/run_manager.go:727]:
```suggestion
			// Validate the credential before adding it to tekton parameters
```

File: pkg/runctl/run_manager_test.go
Comment (alxsap) [inline pkg/runctl/run_manager_test.go:1469]: What do we expect for those values instead?

File: pkg/runctl/run_manager_test.go
Comment (alxsap) [inline pkg/runctl/run_manager_test.go:1652]:
```suggestion
		indexUrlJson string
```

File: pkg/runctl/run_manager_test.go
Comment (alxsap) [inline pkg/runctl/run_manager_test.go:1688]: There should be an error since the URL is wrong

File: test/builder/pipelineRun.go
Comment (alxsap) [inline test/builder/pipelineRun.go:181]: only credential

File: test/builder/pipelineRun.go
Comment (alxsap) [inline test/builder/pipelineRun.go:139]: To be consistent...?
```suggestion
func LoggingWithRunID(runID *api.CustomJSON) PipelineRunSpecOp {
```
Or all of them without `With`.

File: charts/steward/templates/clustertask-jenkinsfile-runner.yaml
Comment (rinckm) [reply charts/steward/templates/clustertask-jenkinsfile-runner.yaml:109]: We must be careful when merging.
We MUST NOT mix 'inputs.params.X' with 'params.Y' variable names.

File: pkg/runctl/run_manager.go
Comment (valiparsa) [reply pkg/runctl/run_manager.go]: Should I add it before function definition of `copyElasticSearchCredentialToRunNamespace`?

File: charts/steward/templates/clustertask-jenkinsfile-runner.yaml
Comment (valiparsa) [reply charts/steward/templates/clustertask-jenkinsfile-runner.yaml]: I only changed it now for `PIPELINE_LOG_ELASTICSEARCH_INDEX_URL` and `PIPELINE_LOG_ELASTICSEARCH_CREDENTIAL`. I guess the rest should be handled by the PR from Andreas #173

File: changelog.yaml
Comment (alxsap) [inline changelog.yaml:24]: We need to change all `>-` to `|-`. Don't touch the existing ones to avoid conflicts.

File: pkg/runctl/run_manager_test.go
Comment (anbrsap) [inline pkg/runctl/run_manager_test.go:1856]: Indentation is wrong (additional space).

File: pkg/runctl/run_manager_test.go
Comment (anbrsap) [inline pkg/runctl/run_manager_test.go:1902]: Indentation is wrong (additional spaces).

File: pkg/runctl/run_manager_test.go
Comment (anbrsap) [inline pkg/runctl/run_manager_test.go:1863]:
```suggestion

			// VERIFY
			assert.ErrorContains(t, err, "field spec.logging.elasticSearch.indexURL is invalid: scheme not supported")
```

File: pkg/runctl/run_manager_test.go
Comment (anbrsap) [inline pkg/runctl/run_manager_test.go:1862]: Don't use the generic name `err` to store the examinee's result, as it may quickly happen that it get's accidently overwritten in the verify block. Better name it `resultError`.

File: pkg/runctl/run_manager_test.go
Comment (anbrsap) [inline pkg/runctl/run_manager_test.go:1915]: Ditto:
```suggestion
	/**
	 * Test: `createTektonTaskRun` handles valid index URLs correctly.
	 */
```

File: pkg/runctl/run_manager_test.go
Comment (anbrsap) [inline pkg/runctl/run_manager_test.go:1925]: You can reduce field `indexURLJson` to just the URL string.

**Review by anbrsap — CHANGES_REQUESTED** [review]

File: pkg/runctl/run_manager.go
Comment (valiparsa) [inline pkg/runctl/run_manager.go:729]:
```suggestion
			elasticSearchURL, err := ensureValidElasticsearchIndexURL(spec.Logging.Elasticsearch.IndexURL)
```

File: pkg/runctl/run_manager_test.go
Comment (valiparsa) [reply pkg/runctl/run_manager_test.go:1932]: done

File: pkg/runctl/run_manager_test.go
Comment (anbrsap) [inline pkg/runctl/run_manager_test.go:1868]:
```suggestion

			// exercise
```

File: pkg/runctl/run_manager_test.go
Comment (anbrsap) [inline pkg/runctl/run_manager_test.go:1885]: Add at least one more test case to justify a parametrized test.

File: pkg/runctl/run_manager_test.go
Comment (anbrsap) [inline pkg/runctl/run_manager_test.go:1965]:
```suggestion

			// verify
			assert.NilError(t, resultError)
```

File: pkg/runctl/run_manager_test.go
Comment (valiparsa) [reply pkg/runctl/run_manager_test.go:1886]: This test will not work anymore because we removed configured IndexURL from tekton parameters here: https://github.com/SAP/stewardci-core/pull/172#discussion_r526677601
I'll remove the corresponding tests for now and we can add them in the future when we add IndexURL in Tekton params

File: pkg/runctl/run_manager.go
Comment (anbrsap) [inline pkg/runctl/run_manager.go:697]:
```suggestion
				return errors.Wrapf(err,
					"field \"spec.logging.elasticsearch.indexURL\" has invalid value %q",
					spec.Logging.Elasticsearch.IndexURL,
				)
```

File: pkg/runctl/run_manager.go
Comment (anbrsap) [inline pkg/runctl/run_manager.go:777]:
```suggestion
	if err != nil {
		return "", err
	}
	if !(strings.ToLower(validURL.Scheme) == "http") && !(strings.ToLower(validURL.Scheme) == "https") {
		return "", fmt.Errorf("scheme not supported: %q", validURL, validURL.Scheme)
	}
```

File: pkg/runctl/run_manager_test.go
Comment (anbrsap) [inline pkg/runctl/run_manager_test.go:1662]: Should be reverted.

File: pkg/runctl/run_manager_test.go
Comment (anbrsap) [inline pkg/runctl/run_manager_test.go:1776]: Can be covered by `Passthrough` test above, can't it?

Comment (valiparsa) [reply pkg/runctl/run_manager_test.go]: Right, It can be. I moved them up

**Review by anbrsap — APPROVED** [review]

**Comment by valiparsa** [comment]
@anbrsap could you please check reviewer's checklist in the PR template before I merge the PR?

**Comment by valiparsa** [comment]
@alxsap @rinckm as there are two approvers required for the PR to be merged that would be helpful if one of you can also review and approve if everything is fine

File: pkg/runctl/run_manager_test.go
Comment (rinckm) [inline pkg/runctl/run_manager_test.go:1824]: This does not match to line 1784
Do you only want to test wrong scheme or also other validation errors?
May be also make error configurable in test.
Add more tests.
"foo///bar"
"foo::bar"

File: test/builder/pipelineRun_test.go
Comment (rinckm) [inline test/builder/pipelineRun_test.go:148]: Looks to me the same as Test_PipelineRunLoggingWithRunIDAndWithIndexURL
May be switch order of 
-LoggingWithIndexURL("testURL"),
- LoggingWithRunID(

to get a new test

File: test/builder/pipelineRun_test.go
Comment (rinckm) [inline test/builder/pipelineRun_test.go:225]:
```suggestion
                       tc := tc
			t.Parallel()
```
Avoid mismatch in test

**Review by rinckm — CHANGES_REQUESTED** [review]
See comments inline

File: pkg/runctl/run_manager_test.go
Comment (valiparsa) [reply pkg/runctl/run_manager_test.go]: I merged two tests together now. We only throw "scheme not supported" error in the run_manager with regards to indexURL so I afraid making it configurable could help. I renamed line 1784 to have better name for the test.

File: test/builder/pipelineRun_test.go
Comment (valiparsa) [reply test/builder/pipelineRun_test.go:121]: You are right. They are the same. I removed the second one.

File: test/builder/pipelineRun_test.go
Comment (valiparsa) [reply test/builder/pipelineRun_test.go:199]: done

**Review by rinckm — APPROVED** [review]

**Comment by valiparsa** [comment]
@anbrsap 
Change has been tested on the loadtest cluster. 
steps to test:
1- Set replica to 0 for the run_controller deployment
2- Build the run_controller locally and start it
3- Start pipelineValidation -> result was green: https://gkecloudciservice.jaas-gcp.cloud.sap.corp/view/pipelineValidation/job/clientPreparationPipeline/99/
4- Start a a pipelineRun object with incorrect indexURL:
```
apiVersion: steward.sap.com/v1alpha1
kind: PipelineRun
metadata:
  generateName: ok-
spec:
  jenkinsFile:
    repoUrl: https://github.com/SAP-samples/stewardci-example-pipelines
    revision: master
    relativePath: success/Jenkinsfile
  logging:
    elasticsearch:
      runID: {"build": 1}
      indexURL: "http//testURL"
```
5- pipelineRun failed with `result: error_config` and error message:
```
message: 'ERROR: preparing failed [PipelineRun{name: ok-97vmp, namespace: steward-test-tenant-50-rnlmb8,
    state: cleaning}]: field "spec.logging.elasticsearch.indexURL" has invalid value
    "http//testURL": scheme not supported: ""'
```

**Pull request merged**
